lib/csv/csv.nit

```diff
@@ -15,119 +15,156 @@
+			if c == delim_char then
+				bf.add c
+			end
+			bf.add c
+		end
+		bf.add '"'
+		return bf.to_s
+	end
+
+	# How many more bytes should be allocated for CSV escaping ?
+	private fun to_escape_csv(sep_char, delim_char: Char, eol: String): Int do
+		var more_ln = 0
+		var ln = length
+		var need_esc = false
+		var fst_eol = eol.first
+		var i = 0
+		while i < ln do
+			var c = self[i]
+			if c == delim_char then more_ln += 1
+			if c == fst_eol then
+				need_esc = true
+				for j in [1 .. eol.length[ do
+					i += 1
+					c = self[i]
+					if c != eol[j] then
+						i -= j
+						need_esc = false
+						break
+					end
+				end
+			end
+			if c == sep_char then need_esc = true
+			i += 1
+		end
+		var more = more_ln * delim_char.u8char_len
+		if need_esc then more += 2
+		return more
+	end
+
+	# Unescape the content of `self` from CSV format to Nit String
+	private fun unescape_csv(delim_char: Char): String do
+		var to_un = to_unescape_csv(delim_char)
+		if to_un == 0 then return to_s
+		var buf = new Buffer.with_cap(bytelen - to_un)
+		var pos = 0
+		var ln = length
+		while pos < ln do
+			var c = self[pos]
+			if c == delim_char then pos += 1
+			buf.add c
+			pos += 1
+		end
+		return buf.to_s
+	end
+
+	# How many bytes should be removed for CSV unescaping ?
+	private fun to_unescape_csv(delim_char: Char): Int do
+		var pos = 0
+		var to_un = 0
+		var ln = length
+		while pos < ln do
+			var c = self[pos]
+			if c == delim_char then
+				pos += 1
+				to_un += 1
+			end
+			pos += 1
+		end
+		return to_un
+	end
+end
+
+# Any class using CSV to write/read data
+class CsvStream
 	# The character that delimits escaped value.
 	#
 	# The delimiter is escaped by doubling it.
-	var delimiter: Char
+	var delimiter = '"' is public writable
 
 	# The character that split each cell in a row.
-	var separator: Char
+	var separator = ',' is public writable
 
 	# The character that ends a row (end of line).
-	var eol: String
-
-	# Escape sequence for the delimiter.
-	private var escaping = "{delimiter}{delimiter}" is lazy
-
-	# Escape the specified cell.
-	private fun escape_cell(cell: String): Text do
-		var result = new RopeBuffer
-		result.add delimiter
-		result.append cell.replace(delimiter, escaping)
-		result.add delimiter
-		return result
-	end
-
-	# Can the specified value be inserted without any escaping?
-	private fun is_value_clean(value: String): Bool do
-		for c in value.chars do
-			if c == delimiter then return false
-			if c == separator then return false
-			if eol.chars.has(c) then return false
-		end
-		return true
-	end
+	var eol = "\n" is public writable
 end
 
 # A CSV document representation.
 class CsvDocument
 	super Writable
-
-	# The format to use.
-	#
-	# Defaults to `rfc4180`.
-	var format: CsvFormat = rfc4180 is writable
+	super CsvStream
 
 	# The header.
 	#
 	# Contains the name of all fields in this table.
-	var header: Array[String] = new Array[String] is writable
+	var header: Array[String] is public writable
```

Comment:
Do you really need the `public` part? If I'm not mistaken, `is writable` should make it public by default.